### PR TITLE
[BugFix] fix unstable sort when using dataloader with HeteroGraph

### PR DESCRIPTION
**Pull request by maqy1995**
## Description
When using Dataloader on a HeteroGraph and setting shuffle to False, the output_nodes are out of order because of the unstable sort. related issue:https://github.com/dmlc/dgl/issues/4125

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## relevant information
related pytorch stable sort issue: https://github.com/pytorch/pytorch/issues/38681
pytorch `torch.sort` doc: https://pytorch.org/docs/stable/generated/torch.argsort.html?highlight=argsort#torch.argsort
pytorch `torch.argsort` doc: https://pytorch.org/docs/stable/generated/torch.argsort.html?highlight=argsort#torch.argsort

## may need attention
According the pytorch doc, the cpu version of `stable torch.sort` was introduced in version 1.9, and the cuda version of `stable torch.sort` was introduced in 1.10.

### CODE REVIEW

**Review by BarclayII — APPROVED** [review]

File: python/dgl/dataloading/dataloader.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def __next__(self):
         # convert the type-ID pairs to dictionary
         type_ids = batch[:, 0]
         indices = batch[:, 1]
-        type_ids_sortidx = torch.argsort(type_ids)
+        _, type_ids_sortidx = torch.sort(type_ids, stable=True)
```
Comment (BarclayII) [inline python/dgl/dataloading/dataloader.py:75]: I guess it's better to raise a warning for older versions of PyTorch when stable sorting is not available and resort to non-stable version (or does PyTorch do that already?)

Comment (maqy1995) [reply python/dgl/dataloading/dataloader.py]: I agree.
It seems older PyTorch will not resort to non-stable version, it just raises an invalid arguments error.
I add the torch version check and raise a warning for older versions of Pytorch. One problem is the warning will also be raised when `shuffle` is `True`, the disorder of the output_nodes should not matter in this scenario.

Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py]: > One problem is the warning will also be raised when `shuffle` is `True`, the disorder of the output_nodes should not matter in this scenario.

Then you could choose to only raise the warning if `shuffle` is False.

> by the way, dgl_warning() seems will repeat the first line?

You mean the DGLWarning will be repeated for each process?  I guess there's not a good way to handle that, but the purpose of warnings is to let users know that there is a problem when `shuffle` is False.
Another solution would be using `np.argsort` (that has a stable option) rather than `torch.sort` if `type_ids` is on CPU, and just raise an error if `type_ids` is on GPU, PyTorch version is too old, and `shuffle=False`.

Comment (maqy1995) [reply python/dgl/dataloading/dataloader.py]: > Then you could choose to only raise the warning if shuffle is False.

I thought about this solution, but there is no shuffle parameter passed in _TensorizedDatasetIter(object).

> You mean the DGLWarning will be repeated for each process?

what I mean here is the first line of the message in dgl_warning will be printed again. For example, below is the full output, the line `The current output_nodes are out of order even if set shuffle to False in Dataloader ` will be printed again at the end of the message.

>DGLWarning: The current output_nodes are out of order even if set shuffle to False in Dataloader, the reason is that the current version of torch dose not support stable sort. Please update torch to 1.10.0 or higher to fix it.
' The current output_nodes are out of order even if set shuffle to False in Dataloader, '
set `line=''` in `_default_formatwarning(message, category, filename, lineno, line=None)`  may avoid this.
```
def dgl_warning_format(message, category, filename, lineno, line=None):
    """Format DGL warnings."""
    if isinstance(category, DGLWarning):
        return "DGL Warning: {}\n".format(message)
    else:
        return _default_formatwarning(message, category, filename, lineno, line=None)
```

> Another solution would be using np.argsort (that has a stable option) rather than torch.sort if type_ids is on CPU, and just raise an error if type_ids is on GPU, PyTorch version is too old, and shuffle=False.

What I was thinking about before was trying to get the same behavior for the cpu or gpu,  and avoid adding too much code.

Comment (BarclayII) [reply python/dgl/dataloading/dataloader.py]: Decided to pass in a shuffle argument to the iterator.

I'm also thinking of the options to just deprecating PyTorch <1.10.  @jermainewang you mentioned previously that we only plan to support the last three stable versions.  Did we formally announce this policy?
(Not an actual review)

**Comment by jermainewang** [comment]
@zjost Could you post your reply to #4224 or create a new thread? It will be hard for people to follow conversations in merged PRs.
